blocklib/picoscope/Picoscope.hpp

```diff
@@ -205,338 +147,404 @@ using A = gr::Annotated<T, description, Arguments...>;
+
+        if (isOpened()) {
             return;
         }
 
         try {
+            open();
             initialize();
-            configure();
-            if (ps_settings.auto_arm) {
+            if (auto_arm) {
                 arm();
             }
-            ps_state.started = true;
+            serial_number = serialNumber();
             fmt::println("Picoscope serial number: {}", serial_number);
         } catch (const std::exception& e) {
             fmt::println(std::cerr, "{}", e.what());
         }
     }
 
     void stop() noexcept {
-        if (!ps_state.started) {
-            return;
-        }
-
-        ps_state.started = false;
-
-        if (!ps_state.initialized) {
-            return;
-        }
-
         disarm();
         close();
     }
 
-    std::string driverVersion() const { return self().driver_driverVersion(); }
-
-    std::string hardwareVersion() const { return self().driver_hardwareVersion(); }
-
-    std::string serialNumber() const { return self().driver_serialNumber(); }
-
-    void initialize() {
-        if (ps_state.initialized) {
+    void disarm() noexcept {
+        if (!isOpened()) {
             return;
         }
 
-        if (const auto ec = self().driver_initialize()) {
-            throw std::runtime_error(fmt::format("Initialization failed: {}", ec.message()));
+        if (const auto status = self().driverStop(_handle); status != PICO_OK) {
+            this->emitErrorMessage(fmt::format("{}::disarm()", this->name), gr::Error(detail::getErrorMessage(status)));
         }
-
-        ps_state.initialized = true;
-
-        serial_number = serialNumber();
+        _isArmed.store(false, std::memory_order_relaxed);
     }
 
-    void close() {
-        self().driver_close();
-        ps_state.closed      = true;
-        ps_state.initialized = false;
-        ps_state.configured  = false;
-    }
-
-    void configure() {
-        if (!ps_state.initialized) {
-            throw std::runtime_error("Cannot configure without initialization");
-        }
-
-        if (ps_state.armed) {
-            throw std::runtime_error("Cannot configure armed device");
+    void open() noexcept {
+        if (isOpened()) {
+            return;
         }
 
-        if (const auto ec = self().driver_configure()) {
-            throw std::runtime_error(fmt::format("Configuration failed: {}", ec.message()));
+        if (const auto status = self().openUnit(serial_number); status == PICO_POWER_SUPPLY_NOT_CONNECTED || status == PICO_USB3_0_DEVICE_NON_USB3_0_PORT) {
+            if (const auto statusPower = self().changePowerSource(_handle, status); statusPower != PICO_OK) {
+                this->emitErrorMessage(fmt::format("{}::open() changePowerSource", this->name), gr::Error(detail::getErrorMessage(statusPower)));
+            }
         }
-
-        ps_state.configured = true;
     }
 
-    void arm() {
-        if (ps_state.armed) {
+    void close() noexcept {
```

Comment:
Why 'noexcept'? This will terminate the program on any even minor/recoverable error.